### DISCUSSION TITLE
Local testing on Avalanche

### DISCUSSION THREAD

**Discussion by HariSeldon23** [dcgtc/dgrants#101] [Ideas]
We would like to test dGrants on Avalanche and then contribute to the codebase as we test and find any issues.

Currently the dApp relies on Infura which only serves Ethereum.

We would like to add support for Avalanche so that our developers can test locally. 

What would be the best approach to do this? I looked for contribution guideline, but couldn't find anything. Should we fork the code and then build those changes in a new repo or should we add multiple L1 support into this repo and submit PR's?

**Comment by thelostone-mc** [comment]
Hey @HariSeldon23 ! 
So dGrants is still in it's early stages and we'd be pushing in changes to get it running for running the first decentralized grant round

If integration is as simple 
- as a building out an adapter which let's you switch one provider for another.
- a few other minor changes

I'd say we should explore having it in the same repo. 

But if it requires a lot of changes (like the contracts themselves), maybe it might be worth 
- forking the repo and making those changes 
- keep rebasing against this until we get to a stable point (this is optional)

Would you be able to get your team to do an analysis/test runs on the level of change needed here 
(simply cause we don't have anyone who's super familiar with avalance)

PS: we are hoping to have a working version (just the minimal backbone) in a month , so if it's not an urgency -> waiting till then and then making a decision might be easier

**Comment by phutchins** [comment]
The ability to test locally is something that I've mentioned would be nice to have. I agree with @thelostone-mc on his approach and think that the answer is likely the first case.

While it is not in scope currently for us to add support for different networks, we absolutely want to make it possible and even easy for others to fork and do so. We should use this as an opportunity to think about what that means and how we can do that.

@HariSeldon23 how would you propose we solve this problem (adding a local testing option)?

**Reply by HariSeldon23** [reply]
Hey everyone, thanks for the comments.

So there's two main levels of change's we'll look at doing:
* Smart contract related functionality
* Front end

# Smart contracts
We won't be changing any logic or anything key here. There may be a few small changes required here to do with how blocks are produced in Ethereum versus Avalanche. Generally these relate to the Timestamp calculations. So one potential approach would be to have a prefix to all Solidity files such as
* GrantRounde (to indicate Ethereum)
* GrantRounda (to indicate Avalanche)

Then the front end could then pick up the network it's pointing at and route accordingly.

# Front end
This is where the vast majority of the work we want to do will occur.

Firstly the `env` file points to Infura. There's no Infura on Avalanche so the first step would be to allow us to point the dApp at either a local/testnet/mainnet node to be used. We could then also add a flag to the `node` commands which would specify the network being used.

Most of our changes wouldn't happen in the code base at this stage. Our workflow for building dApps (excluding the Smart Contracts) normally follows this process:
UX Designer builds low fidelity prototypes -> Release prototypes to a research group -> Incorporate the changes into a high fidelity prototype -> Release prototype to broader audience for testing -> Sign off prototypes -> Build the dApp.

So my current thinking is that we'll assign our UX designer to start working on the low fidelity prototypes.
